Modified the test result directory structure naming convention and more (#2030)
oracle/weblogic-kubernetes-operator: integration-tests/README.md

- Install supported version of JDK. Need JDK version 11+ or later.
- Clone the `weblogic-kubernetes-operator` repository.
- Set up a personal account on `container-registry.oracle.com`.
- Log in to `https://container-registry.oracle.com` and accept Oracle Standard Terms and Restrictions to pull WebLogic images from `container-registry.oracle.com/middleware/weblogic`.
- Export the following environment variables before running the tests:
   ```
    export BASE_IMAGES_REPO="container-registry.oracle.com"
    export OCR_EMAIL=<ocr_email>
    export OCR_USERNAME=<ocr_username>
    export OCR_PASSWORD=<ocr_password> 
   ```
- `cd  weblogic-kubernetes-operator` and run the following commands:
  ```
    # Build the WebLogic Server Kubernetes Operator binary 
    mvn clean install 
    # Run tests in sequential order
    mvn -pl -integration-tests -P integration-tests clean verify
    # Run tests in parallel
    mvn -DPARALLEL=true -DNUMBER_OF_THREADS=2 -pl -integration-tests -P integration-tests verify
   ```
## How to run individual test class(es)

> mvn -Dit.test=ItMiiDomain,ItMiiUpdateDomainConfig -pl integration-tests -P integration-tests verify

Reviewer: Shouldn't these test commands (this section and the next two) be in code font?
```
mvn -Dit.test=ItMiiDomain,ItMiiUpdateDomainConfig -pl integration-tests -P integration-tests verify
```

# Integration tests for Oracle WebLogic Server Kubernetes Operator

> This documentation describes the procedure to run integration tests for the WebLogic Server Kubernetes Operator locally in the Oracle Linux environment. The tests are written in the JUnit5 framework and driven by maven profile.

Author: done

# Directory Configuration and Structure
Directory structure of integration test source is distributed as follows 
```
weblogic-kubernetes-operator/integration-tests - location of module pom.xml  
weblogic-kubernetes-operator/integration-tests/src/test/java/oracle/weblogic/kubernetes - integration test(JUnit5) classes and utility classes  
weblogic-kubernetes-operator/integration-tests/src/test/resources - properties, YAML files and other bash scripts
```

> # How does integration test infrastructure works

Reviewer 2: ```suggestion
# How does Operator integration tests work?
```

Author: done

>  - Build weblogic kubernetes operator image from the downloaded branch.

Author: done

 - Download the latest version of WebLogic Deploy Tooling (WDT) and WebLogic Image Tool (WIT)
 - Pull the specified version of WebLogic Image from container-registry.oracle.com (OCR) or phx.ocir.io (OCIR) 

>  - Build a simple weblogic domain image with a dynamic weblogic cluster and a sample web application using Model-In-Image model.

Reviewer 3: ```suggestion
 - Build a simple WebLogic domain in image with a dynamic WebLogic cluster and a sample web application using Model-In-Image model.
```

>  - Build a simple WebLogic domain in image with a dynamic WebLogic cluster and a sample web application using Model-In-Image model.

Reviewer 2: - Create domain resource and make sure the WebLogic domain is up and running.

Author: done

 - Install istio service mesh.

>  - After test execution, clean all the Namespaces, Kubernetes Objects created during test execution.

Reviewer 3: ```suggestion
 - After test execution, delete all the Namespaces, Kubernetes Objects created during test execution.
```

Author: done

>  - Archive the test stdout for each test class and kubernates object details for each test method in the diagnostic directory for triage. 

Reviewer 3: ```suggestion
 - Archive the test stdout for each test class and kubernates object details for each test method in the diagnostic directory for triage when tests fail or when the env variable COLLECT_LOGS_ON_SUCCESS=true is set
```

Author: done

>  - Archive the test stdout for each test class and kubernates object details for each test method in the diagnostic directory for triage when tests fail or when the env variable COLLECT_LOGS_ON_SUCCESS is set to true.

Reviewer 2: ```suggestion
 - Archive the test stdout for each test class and kubernetes object details for each test method in the diagnostic directory for triage when tests fail or when the env variable COLLECT_LOGS_ON_SUCCESS is set to true.
```

Author: done

 
# How to run Operator integration tests locally on Oracle Linux

> - Install supported version Helm and Kubernetes cluster. For detail see [Weblogic Kubernetes Operator guide](https://oracle.github.io/weblogic-kubernetes-operator/userguide/introduction/introduction/).

Reviewer 3: ```suggestion
- Install supported version of Helm and Kubernetes cluster. For detail see [Weblogic Kubernetes Operator guide](https://oracle.github.io/weblogic-kubernetes-operator/userguide/introduction/introduction/).
```

Author: done

> - Install supported version JDK. Need JDK Version 11+ and above.

Reviewer 3: ```suggestion
- Install supported version of JDK. Need JDK Version 11+ and above.
```

Author: done

> - Install supported version JDK. Need JDK Version 11+ and above.

Reviewer 3: - Install maven (if any version need to be specified include it here)

- Clone the weblogic-kubernetes-operator repository
- Setup a personal account on container-registry.oracle.com

> - Log into container-registry.oracle.com and signup for access to WebLogic 12.2.1.4 images from container-registry.oracle.com/middleware/weblogic:12.2.1.4

Reviewer 3: ```suggestion
- Log into https://container-registry.oracle.com and accept Oracle Standard Terms and Restrictions  to pull WebLogic 12.2.1.4 images from container-registry.oracle.com/middleware/weblogic:12.2.1.4
```

Author: done

- Export the following environment variable before running the tests
   ```
    export BASE_IMAGES_REPO="container-registry.oracle.com"
    export OCR_EMAIL=<ocr_email>
    export OCR_USERNAME=<ocr_username>
    export OCR_PASSWORD=<ocr_password> 
   ```

> - In weblogic-kubernetes-operator run the following commands

Reviewer 3: ```suggestion
- cd weblogic-kubernetes-operator and run the following commands
```

Author: done

  ```
    # Build the Weblogic Kubernetes Operator binary 
    mvn clean install 
    # Run tests in sequential

>     mvn -pl -integration-tests -P integration-tests verify

Reviewer 3: ```suggestion
    mvn -pl -integration-tests -P integration-tests clean verify
```

Author: done

    # Run tests in parallel
    mvn -DPARALLEL=true -DNUMBER_OF_THREADS=2 -pl -integration-tests -P integration-tests verify
   ```
## How to run individual test class(es)
mvn -Dit.test=ItMiiDomain,ItMiiUpdateDomainConfig -pl integration-tests -P integration-tests verify

## How to run a single test method in a test class 
mvn -Dit.test="ItMiiUpdateDomainConfig#testMiiDeleteSystemResources" -pl integration-tests -P integration-tests verify

## How to exclude test class(es) 
mvn -Dit.test="!ItCrossDomainTransaction,!ItMiiUpdateDomainConfig" -pl integration-tests -P integration-tests verify

## Environment variables to manage test execution 
| Variable | Description | Default Value
| --- | --- | --- |

> | SKIP_CLEANUP  | Test infra remove for all kubernetes objects created during test execution. To retain all such objects for triaging a test failure this environment variable should be set to true. User need to run weblogic-kubernetes-operator/src/integration-tests/bash/cleanup.sh to clean up kubernetes objects later.    | true

Reviewer 3: ```suggestion
| SKIP_CLEANUP  | Test infrastructure removes all kubernetes objects created during test execution. To retain all such objects for triaging test failures this environment variable should be set to true. User need to run weblogic-kubernetes-operator/src/integration-tests/bash/cleanup.sh to clean up kubernetes objects after traiging.    | true
```

Author: done

> | SKIP_CLEANUP  | Test infrastructure removes all kubernetes objects created during test execution. To retain all such objects for triaging test failures this environment variable should be set to true. User need to run weblogic-kubernetes-operator/src/integration-tests/bash/cleanup.sh to clean up kubernetes objects after traiging.    | true

Reviewer 2: Default is false...

Author: done

| COLLECT_LOGS_ON_SUCCESS  | Test infra does not keep the diagnostic log for successful tests. To archive the diagnostic log for successful tests, this environment variable should be set to true. | false

> | LOGS_DIR  | Root directory for the diagnostic logs. | /tmp/it-diagnosticlogs

Reviewer 3: Please see my other comment for the BASE_DIR for test logs

Author: done

# Integration tests for Oracle WebLogic Server Kubernetes Operator

> This documentation describes the steps to run integration tests for the Oracle WebLogic Server Kubernetes Operator locally on a Oracle Linux environment. The tests are written in the JUnit5 framework and driven by maven profile.

Author: done

# Directory Configuration and Structure

> Directory structure of integration test source is distributed as follows 

Author: done

```
weblogic-kubernetes-operator/integration-tests - location of module pom.xml  
weblogic-kubernetes-operator/integration-tests/src/test/java/oracle/weblogic/kubernetes - integration test(JUnit5) classes and utility classes  
weblogic-kubernetes-operator/integration-tests/src/test/resources - properties, YAML files and other bash scripts
```

> # How does WebLogic Kubernetes Operator integration test work

Author: done

> # How does WebLogic Kubernetes Operator integration test work

Reviewer: Not done. Choices are: How do WebLogic Server Kubernetes Operator integration tests work? OR How does the WebLogic Server Kubernetes Operator integration tests work?

> # How does WebLogic Kubernetes Operator integration test work

Author: [modified] # How do WebLogic Server Kubernetes Operator integration tests work.

Should I put a period (.)  or ?  at the end even if this is a heading.  I have similar headings but I do not put ? at the end.
e.g.  # How to run operator integration tests locally on Oracle Linux. 

What is your suggestion ?

> # How does WebLogic Kubernetes Operator integration test work

Reviewer: When you have, "How do WebLogic Server Kubernetes Operator integration tests work," it asks a question, therefore, needs a question mark. -> How do WebLogic Server Kubernetes Operator integration tests work?

The other section titles, such as, "How to run operator integration tests locally on Oracle Linux." do not need a period after them.

> # How does WebLogic Kubernetes Operator integration test work

Author: Modified the heading as follows ...
 # How do WebLogic Server Kubernetes Operator integration tests work?
 # How to run operator integration tests locally on Oracle Linux

> # How does WebLogic Kubernetes Operator integration test works

Reviewer 2: ```suggestion
# How does WebLogic Kubernetes Operator integration test work
```

Author: done

>  - Build WebLogic Kubernetes Operator image from the downloaded git branch.

Author: done

>  - Install the latest version of WebLogic Deploy Tooling (WDT) and WebLogic Image Tool (WIT)

Author: done

>  - Install supported istio service mesh.

Author: done

>  - Pull the specified version of WebLogic image from container-registry.oracle.com (OCR)

Author: done

>  - Pull the specified version of WebLogic Image from container-registry.oracle.com (OCR)

Reviewer 2: ```suggestion
 - Pull the specified version of WebLogic image from container-registry.oracle.com (OCR)
```

>  - Install WebLogic Kubernetes Operator.

Reviewer: WebLogic Kubernetes Operator -> WebLogic Server Kubernetes Operator

Author: done

>  - Build a WebLogic domain resource WebLogic.

Reviewer 2: ```suggestion
 - Build a WebLogic domain resource.
```

Author: done

>  - Make sure WebLogic domain is running by verifying the corresponding server pod and kubernetes service status. 

Reviewer: sure WebLogic domain -> sure that the WebLogic domain
kubernetes  -> Kubernetes

Author: done

>  - Make sure WebLogic domain is running by verify the corresponding server pod and kubernetes service status. 

Reviewer 2: ```suggestion
 - Make sure WebLogic domain is running by verifying the corresponding server pod and kubernetes service status. 
```

Author: done

>  - After test execution, delete all the Namespaces, Kubernetes Objects created during test execution.

Reviewer: Namespaces, Kubernetes Objects -> namespaces and Kubernetes objects

Author: done

>  - Archive the test standard output for each test class and kubernetes object details for each test method in the diagnostic directory for triage when tests fail or when the env variable COLLECT_LOGS_ON_SUCCESS is set to true.

Reviewer: kubernetes  -> Kubernetes 
COLLECT_LOGS_ON_SUCCESS -> `COLLECT_LOGS_ON_SUCCESS` (surround with back ticks to give it code font)

Author: done

> # How to run Operator integration tests locally on Oracle Linux

Reviewer: Operator  -> operator 
Linux -> Linux. (period)

Author: done

> - Install supported version of Helm and Kubernetes cluster. For detail see [Weblogic Kubernetes Operator guide](https://oracle.github.io/weblogic-kubernetes-operator/userguide/introduction/introduction/).

Reviewer: For detail see [Weblogic Kubernetes Operator guide] -> For details, see the [WebLogic Server Kubernetes Operator guide]

Author: done

Reviewer: Not done. For details, see the [WebLogic Server Kubernetes Operator guide]

Author: See the link  https://github.com/oracle/weblogic-kubernetes-operator/blob/6719d05560ec120faeaf951ca9354543e6449491/integration-tests/README.md#how-to-run-operator-integration-tests-locally-on-oracle-linux

Author: [modified] 
Install supported version of Helm and Kubernetes cluster. For details, see the [WebLogic Server Kubernetes Operator guide](https://oracle.github.io/weblogic-kubernetes-operator/userguide/introduction/introduction/).

> - Install supported version of JDK. Need JDK Version 11+ and above.

Reviewer: JDK Version -> JDK version
and above. -> or later.

Author: done

> - Clone the weblogic-kubernetes-operator repository

Reviewer: weblogic-kubernetes-operator -> `weblogic-kubernetes-operator` (surround with back ticks to give it code font)

Author: done

> - Setup a personal account on container-registry.oracle.com

Reviewer: Setup  -> Set up 
container-registry.oracle.com -> `container-registry.oracle.com`. (surround with back ticks to give it code font) (period at the end of the sentence)

Author: done

> - Log into https://container-registry.oracle.com and accept Oracle Standard Terms and Restrictions to pull WebLogic images from container-registry.oracle.com/middleware/weblogic

Reviewer: Log into -> Log in to
https://container-registry.oracle.com -> `https://container-registry.oracle.com` (surround with back ticks to give it code font)
container-registry.oracle.com/middleware/weblogic -> `container-registry.oracle.com/middleware/weblogic`. (surround with back ticks to give it code font) (period at the end of the sentence)

Author: done

Reviewer: Not done. Log in to

Author: [modified]  Log in to `https://container-registry.oracle.com` and accept Oracle Standard Terms and Restrictions to pull WebLogic images from `container-registry.oracle.com/middleware/weblogic`.

> - Export the following environment variable before running the tests

Reviewer: variable  -> variables
tests -> tests: (colon)

Author: done

   ```
    export BASE_IMAGES_REPO="container-registry.oracle.com"
    export OCR_EMAIL=<ocr_email>
    export OCR_USERNAME=<ocr_username>
    export OCR_PASSWORD=<ocr_password> 
   ```

> - `cd  weblogic-kubernetes-operator` and run the following commands

Reviewer: commands -> commands: (colon)

Author: done

  ```

>     # Build the Weblogic Kubernetes Operator binary 

Reviewer: Weblogic Kubernetes Operator -> WebLogic Server Kubernetes Operator

Author: done

    mvn clean install 

>     # Run tests in sequential

Reviewer: in sequential -> in sequential order

Author: done

    mvn -pl -integration-tests -P integration-tests clean verify
    # Run tests in parallel
    mvn -DPARALLEL=true -DNUMBER_OF_THREADS=2 -pl -integration-tests -P integration-tests verify
   ```
## How to run individual test class(es)
mvn -Dit.test=ItMiiDomain,ItMiiUpdateDomainConfig -pl integration-tests -P integration-tests verify

## How to run a single test method in a test class 
mvn -Dit.test="ItMiiUpdateDomainConfig#testMiiDeleteSystemResources" -pl integration-tests -P integration-tests verify

## How to exclude test class(es) 
mvn -Dit.test="!ItCrossDomainTransaction,!ItMiiUpdateDomainConfig" -pl integration-tests -P integration-tests verify

## Environment variables to manage test execution 
| Variable | Description | Default Value
| --- | --- | --- |

> | SKIP_CLEANUP  | Test infrastructure removes all kubernetes objects created during test execution. To retain all such objects for triaging test failures this environment variable should be set to true. User need to run weblogic-kubernetes-operator/src/integration-tests/bash/cleanup.sh to clean up kubernetes objects after traiging.    | false

Reviewer: SKIP_CLEANUP ->`SKIP_CLEANUP` (surround with back ticks to give it code font. Please do this for every variable name!)
kubernetes  -> Kubernetes 
failures -> failures, (comma)
User  -> Users
weblogic-kubernetes-operator/src/integration-tests/bash/cleanup.sh -> `weblogic-kubernetes-operator/src/integration-tests/bash/cleanup.sh` (code font)
kubernetes  -> Kubernetes (please capitalize Kubernetes everywhere in the README)

Author: done

> | COLLECT_LOGS_ON_SUCCESS  | Test infra does not keep the diagnostic log for successful tests. To archive the diagnostic log for successful tests, this environment variable should be set to true. | false

Reviewer: infra  -> infrastructure

Author: done

> | RESULT_ROOT  | Root directory for the integration test results and artifacts generated during test execution. | /tmp/it-testsresults

Reviewer: /tmp/it-testsresults -> `/tmp/it-testsresults` (code font)

Author: done

> | BASE_IMAGES_REPO  | The repository URL to download WebLogic and FMW image. Make sure you have access to this repository. Other supported repository is phx.ocir.io (OCIR)  | container-registry.oracle.com(OCR)

Reviewer: Other  -> The other 
phx.ocir.io -> `phx.ocir.io` (code font)
(OCIR) -> (OCIR). (period)
container-registry.oracle.com(OCR) -> container-registry.oracle.com (OCR) (need a space between registry name and acronym)

Author: done

Reviewer: Not done.  (OCIR). (period)

> | WEBLOGIC_IMAGE_NAME  | Name of the WebLogic image in the chosen repository.| middleware/weblogic (OCR) 

Reviewer: middleware/weblogic -> `middleware/weblogic` (code font)

Author: done

> | WEBLOGIC_IMAGE_TAG  | The tag for WebLogic base image. Generally, it represents the WebLogic version with JDK and/or installation type. Possible values are 12.2.1.3, 12.2.1.3-dev, 12.2.1.4-slim, 14.1.1.0-11 or 14.1.1.0-8. Please check the repository for the availability of these images. | 12.2.1.4

Reviewer: WebLogic version -> WebLogic Server version

Author: done

| FMWINFRA_IMAGE_TAG  | The tag for Fusion Middleware Infrastructure base image. Generally, it represents the FMW version. Possible values are 12.2.1.3, 12.2.1.4. Please check the repository for the availability of these images. | 12.2.1.4

> | FMWINFRA_IMAGE_NAME  | Name of the Fusion Middleware Infrastructure image in the chosen repository.| middleware/fmw-infrastructure (OCR) 

Reviewer: middleware/fmw-infrastructure -> `middleware/fmw-infrastructure` (code font)

Author: done

## Logging/Archiving

> On completion of integration test execution, the results are archived in a directory based on environment variable  LOG_DIR  ( the default value is /tmp/it-diagnosticlogs) and runtime artifacts are available in a directory based on environment variable RESULTS_ROOT ( the default value is /tmp/it-testresults) . 

Reviewer 2: ```suggestion
On completion of integration test execution, the results are archived in a directory based on environment variable  LOG_DIR  ( the default value is /tmp/it-testsresults/diagnostics) and runtime artifacts are available in a directory based on environment variable RESULTS_ROOT ( the default value is /tmp/it-testresults) . 
```

Author: done

> On completion of integration test execution, the results are archived in a directory based on environment variable  LOG_DIR  (the default value is /tmp/it-testsresults/it-diagnostic) and runtime artifacts are available in a directory based on environment variable RESULTS_ROOT (the default value is /tmp/it-testresults) . 

Reviewer: On completion of integration test execution, -> After the completion of running the integration tests,
directory based -> directory-based
LOG_DIR -> `LOG_DIR` (code font)
/tmp/it-testsresults/it-diagnostic -> `/tmp/it-testsresults/it-diagnostic` (code font)
RESULTS_ROOT -> `RESULTS_ROOT` (code font)
/tmp/it-testresults -> `/tmp/it-testresults` (code font)

Author: done

Reviewer: Slight confusion here. Choices are: a directory-based environment variable OR a directory based on the environment variable (I think the latter is what you want)

Author: [modified]  After the completion of running the integration tests, the results are archived in a directory-based environment variable

Reviewer: How about this one: and runtime artifacts are available in a directory based on environment variable
Should it be: and runtime artifacts are available in a directory-based environment variable

Author: [ Modified ] After the completion of running the integration tests, the results are archived in a directory-based environment variable `LOG_DIR` (the default value is `/tmp/it-testsresults/diagnostic`) and runtime artifacts are available in a directory-based environment variable `RESULTS_ROOT` (the default value is `/tmp/it-testresults`).

> A typical diagnosticlogs directory content will look as follows after the completion of test ItMiiDomain. The standard output for the tests is captured in ItMiiDomain.out. For each test method ( say testCreateMiiDomain, testCreateMiiSecondDomain) a directory is created and the corresponding k8s object description log(s) and server pod logs are saved if the test fails or environment variable COLLECT_LOGS_ON_SUCCESS set to true.

Reviewer: A typical diagnosticlogs directory content will look as follows after the completion of test ItMiiDomain. -> After the completion of the test `ItMiiDomain`, a typical diagnostic logs directory content will look as follows.
ItMiiDomain.out -> `ItMiiDomain.out`
(say testCreateMiiDomain, testCreateMiiSecondDomain) -> (for example, `testCreateMiiDomain`, `testCreateMiiSecondDomain`), (comma)
k8s -> Kubernetes
COLLECT_LOGS_ON_SUCCESS -> `COLLECT_LOGS_ON_SUCCESS` (code font)

Author: done

```
itMiiDomain

|-- ItMiiDomain.out

|-- testCreateMiiDomain

|   |-- ns-gggs.list.services.log

|   |-- ns-gggs.pod.weblogic-operator-7b8-lwqzm.container.weblogic-operator.log

|   |-- ns-luaf.pod.domain1-admin-server.container.weblogic-server.log

|   |-- ns-luaf.pod.domain1-managed-server1.container.weblogic-server.log

|   |-- ns-luaf.pod.domain1-managed-server2.container.weblogic-server.log

-- testCreateMiiSecondDomain

    |-- ns-gggs.list.configmaps.log

    |-- ns-gggs.list.domains.log

    |-- ns-gggs.list.services.log

    |-- ns-gggs.pod.weblogic-operator-7b8f-lwqzm.container.weblogic-operator.log

    |-- ns-luaf.list.configmaps.log

    |-- ns-luaf.pod.domain1-admin-server.container.weblogic-server.log

    |-- ns-luaf.pod.domain1-managed-server1.container.weblogic-server.log
```
## Troubleshooting

> The diagnostic logs files in method scoped directory in ${LOGS_DIR}/`<testClass>`/`<testMehod>` and the test stdout ${LOGS_DIR}/`<testClass>`.out is the starting point to triage a test failure.

Reviewer: in method scoped directory -> in the method-scoped directory

Author: done
